File: SproInstall/SocketPro/tutorial/CSharp/SampleThree/Client/TThree.cs


using System;
using SocketProAdapter;
using SocketProAdapter.ClientSide;
using USOCKETLib;
using SampleThreeShared;
using System.Collections;

public class CTThree : CAsyncServiceHandler
{
    public CTThree(CClientSocket cs, IAsyncResultsHandler pDefaultAsyncResultsHandler)
        : base(TThreeConst.sidCTThree, cs, pDefaultAsyncResultsHandler)
    {
    }

    public CTThree(CClientSocket cs)
        : base(TThreeConst.sidCTThree, cs)
    {
    }

    public CTThree()
        : base(TThreeConst.sidCTThree)
    {
    }

	public CTestItem GetOneItem()
	{
        CTestItem ti;
        bool bProcessRy = ProcessR1(TThreeConst.idGetOneItemCTThree, out ti);
        return ti;
	}

	public void SendOneItem(CTestItem Item)
	{
        bool bProcessRy = ProcessR0(TThreeConst.idSendOneItemCTThree, Item);
	}

    public Stack GetManyItems(int nCount)
	{
        m_Stack.Clear();
        bool bProcessR0 = ProcessR0(TThreeConst.idGetManyItemsCTThree, nCount);
        return m_Stack;
	}

    public void SendManyItems(Stack outStack)
    {
        m_OutStack = outStack;
        SendBatchItems();
        WaitAll();
        //at last, we inform a server that we finish sending items
        bool bProcessRy = ProcessR0(TThreeConst.idSendManyItemsCTThree);
    }

    private void SendBatchItems()
    {
        int nSndSize;
        const int BATCH_SIZE_IN_BYTE = 40 * 1024;
        const int BATCH_ITEM_COUNT = 200;
        int nBatch = 0;
        using (CScopeUQueue su = new CScopeUQueue())
        {
            CUQueue UQueue = su.UQueue;
            UQueue.SetSize(0);
            nSndSize = GetAttachedClientSocket().GetUSocket().BytesInSndMemory;
            while (nBatch < BATCH_ITEM_COUNT && nSndSize < BATCH_SIZE_IN_BYTE && m_OutStack != null && m_OutStack.Count > 0)
            {
                CTestItem Item = (CTestItem)m_OutStack.Pop();
                UQueue.Push(Item);
                nBatch++;
                if (nBatch == BATCH_ITEM_COUNT)
                {
                    SendRequest(TThreeConst.idSendBatchItemsCTThree, UQueue);
                    UQueue.SetSize(0);
                    nSndSize = GetAttachedClientSocket().GetUSocket().BytesInSndMemory;
                    if (nSndSize < BATCH_SIZE_IN_BYTE)
                        nBatch = 0;
                    else
                        break;
                }
            }
            if (UQueue.GetSize() > 0)
            {
                SendRequest(TThreeConst.idSendBatchItemsCTThree, UQueue);
                UQueue.SetSize(0);
            }
        }
    }

    protected override void OnResultReturned(short sRequestID, CUQueue UQueue)
    {
        switch (sRequestID)
        {
            case TThreeConst.idGetBatchItemsCTThree:
                while (UQueue.GetSize() > 0)
                {
                    CTestItem Item;
                    UQueue.Pop(out Item);
                    m_Stack.Push(Item);
                }
                break;
            case TThreeConst.idSendBatchItemsCTThree:
                SendBatchItems();
                break;
            default:
                break;
        }
    }

    private Stack m_OutStack;
    private Stack m_Stack = new Stack();
}

File: SproInstall/SocketPro/tutorial/VBNet/SampleThree/Client/TThree.vb
Imports Microsoft.VisualBasic
Imports System
Imports SocketProAdapter
Imports SocketProAdapter.ClientSide
Imports USOCKETLib
Imports SampleThreeShared
Imports System.Collections

Public Class CTThree : Inherits CAsyncServiceHandler
    Public Sub New()
        MyBase.New(TThreeConst.sidCTThree)
    End Sub

    Public Sub New(ByVal cs As CClientSocket)
        MyBase.New(TThreeConst.sidCTThree, cs)
    End Sub
    Public Sub New(ByVal cs As CClientSocket, ByVal DefaultAsyncResultsHandler As IAsyncResultsHandler)
        MyBase.New(TThreeConst.sidCTThree, cs, DefaultAsyncResultsHandler)
    End Sub

    Public Function GetOneItem() As CTestItem
        Dim ti As CTestItem = Nothing
        Dim b As Boolean = ProcessR1(TThreeConst.idGetOneItemCTThree, ti)
        Return ti
    End Function

    Public Sub SendOneItem(ByVal Item As CTestItem)
        Dim b As Boolean = ProcessR0(TThreeConst.idSendOneItemCTThree, Item)
    End Sub

    Public Function GetManyItems(ByVal nCount As Integer) As Stack
        m_Stack.Clear()
        Dim b As Boolean = ProcessR0(TThreeConst.idGetManyItemsCTThree, nCount)
        Return m_Stack
    End Function
    Public Sub SendManyItems(ByVal outStack As Stack)
        m_OutStack = outStack
        SendBatchItems()
        WaitAll()
        'at last, we inform a server that we finish sending items
        Dim b As Boolean = ProcessR0(TThreeConst.idSendManyItemsCTThree)
    End Sub

    Private Sub SendBatchItems()
        Dim nBatch As Integer = 0
        Dim nSndSize As Integer
        Dim UQueue As CUQueue = New CScopeUQueue().UQueue
        UQueue.SetSize(0)
        nSndSize = GetAttachedClientSocket().GetUSocket().BytesInSndMemory
        Do While nBatch < 200 AndAlso nSndSize < 40960 AndAlso Not m_OutStack Is Nothing AndAlso m_OutStack.Count > 0
            Dim Item As CTestItem = CType(m_OutStack.Pop(), CTestItem)
            UQueue.Push(Item)
            nBatch += 1
            If (nBatch = 200) Then
                SendRequest(TThreeConst.idSendBatchItemsCTThree, UQueue)
                UQueue.SetSize(0)
                nSndSize = GetAttachedClientSocket().GetUSocket().BytesInSndMemory
                If (nSndSize < 40960) Then
                    nBatch = 0
                Else
                    Exit Do
                End If
            End If
        Loop
        If UQueue.GetSize() > 0 Then
            SendRequest(TThreeConst.idSendBatchItemsCTThree, UQueue)
            UQueue.SetSize(0)
        End If
    End Sub
    Private m_OutStack As Stack
    Private m_Stack As Stack = New Stack()

    'We can process returning results inside the function.
    Protected Overrides Sub OnResultReturned(ByVal sRequestID As Short, ByVal UQueue As CUQueue)
        Select Case sRequestID
            Case TThreeConst.idGetBatchItemsCTThree
                Do While UQueue.GetSize() > 0
                    Dim Item As CTestItem = Nothing
                    UQueue.Pop(Item)
                    m_Stack.Push(Item)
                Loop
            Case TThreeConst.idSendBatchItemsCTThree
                SendBatchItems()
            Case Else
        End Select
    End Sub
End Class
